lib/ansible/modules/tacp/tacp_instance.py

```diff
@@ -80,278 +78,727 @@
+    except tacp_exceptions.InvalidDiskSizeException:
+        failure_reason = 'Could not add disk to instance; disks must have a positive size in GB provided.'  # noqa
 
-        result['changed'] = True
-
-    # Current state is first dimension
-    # Specified state is second dimension
-    power_state_dict = {
-        (State.RUNNING, Action.STARTED): [],
-        (State.RUNNING, Action.SHUTDOWN): [Action.SHUTDOWN],
-        (State.RUNNING, Action.STOPPED): [Action.STOPPED],
-        (State.RUNNING, Action.RESTARTED): [Action.RESTARTED],
-        (State.RUNNING, Action.FORCE_RESTARTED): [Action.FORCE_RESTARTED],
-        (State.RUNNING, Action.PAUSED): [Action.PAUSED],
-        (State.RUNNING, Action.ABSENT): [Action.ABSENT],
-        (State.SHUTDOWN, Action.STARTED): [Action.STARTED],
-        (State.SHUTDOWN, Action.SHUTDOWN): [],
-        (State.SHUTDOWN, Action.STOPPED): [],
-        (State.SHUTDOWN, Action.RESTARTED): [Action.STARTED],
-        (State.SHUTDOWN, Action.FORCE_RESTARTED): [Action.STARTED],
-        (State.SHUTDOWN, Action.PAUSED): [Action.STARTED, Action.PAUSED],
-        (State.SHUTDOWN, Action.ABSENT): [Action.ABSENT],
-        (State.PAUSED, Action.STARTED): [Action.RESUMED],
-        (State.PAUSED, Action.SHUTDOWN): [Action.RESUMED, Action.SHUTDOWN],
-        (State.PAUSED, Action.STOPPED): [Action.STOPPED],
-        (State.PAUSED, Action.RESTARTED): [Action.RESUMED, Action.RESTARTED],
-        (State.PAUSED, Action.FORCE_RESTARTED): [Action.RESUMED, Action.FORCE_RESTARTED],
-        (State.PAUSED, Action.PAUSED): [],
-        (State.PAUSED, Action.ABSENT): [Action.ABSENT]
-    }
+    if failure_reason:
+        fail_and_rollback_instance_creation()
+
+    RESOURCES['update_app'].create_disk(body=disk_payload, uuid=instance_uuid)
+
+
+def get_disk_payload(playbook_disk):
+    """Generates a payload for creating a new disk in an application.
+
+    Args:
+        playbook_disk (dict): The configuration of a single disk from the
+            Ansible playbook
 
-    # Return the inputs for debugging purposes
-    result['args'] = module.params
+    Raises:
+        tacp_exceptions.InvalidDiskBandwidthLimitException
+        tacp_exceptions.InvalidDiskIopsLimitException
+        tacp_exceptions.InvalidDiskSizeException
+        tacp_exceptions.InvalidDiskNameException
 
-    # Define configuration
-    configuration = tacp_utils.get_configuration(module.params['api_key'],
-                                                 module.params['portal_url'])
-    api_client = tacp.ApiClient(configuration)
+    Returns:
+        ApiDiskSizeAndLimitPayload: The populated payload object to be provided
+            to the function that actually creates the disk.
+    """
+    bandwidth_limit = playbook_disk.get('bandwidth_limit')
+    if bandwidth_limit:
+        if int(bandwidth_limit) < MINIMUM_BW_FIVE_MBPS_IN_BYTES:
+            raise tacp_exceptions.InvalidDiskBandwidthLimitException
```

Comment:
I've always raised with an instance. One benefit is that you can add the message here to clean up the code.
```python
raise tacp_exceptions.InvalidDiskBandwidthLimitException(
    'Could not add disk to instance; disks must have a bandwidth limit of at least {} MBps ({}).'.format(
        ... use constants
    )
)
```

Then above:
```python
try:
    disk_payload = get_disk_payload(playbook_disk)
except Exception as e:
    fail_and_rollback_instance_creation(str(e), instance_uuid)
```